src/mam4xx/aero_model.hpp

```diff
@@ -810,6 +835,171 @@ bool examine_prec_exist(const int level_for_precipitation, const Real pdel[],
+  const int lcoarnacl = aerosol_index_for_mode(ModeIndex::Coarse, AeroId::NaCl);
+  const Real tmpdust =
+      haero::max(0.0, state_q(kk, lcoardust) + ptend_q(kk, lcoardust) * dt);
+  const Real tmpnacl =
+      haero::max(0.0, state_q(kk, lcoarnacl) + ptend_q(kk, lcoarnacl) * dt);
+  if (tmpdust + tmpnacl > small_value_30)
+    f_act_conv_coarse =
+        (f_act_conv_coarse_dust * tmpdust + f_act_conv_coarse_nacl * tmpnacl) /
+        (tmpdust + tmpnacl);
+}
+
+// =============================================================================
+KOKKOS_INLINE_FUNCTION
+void calc_resusp_to_coarse(const int mm, const bool update_dqdt,
+                           const Real rcscavt, const Real rsscavt,
+                           Real &dqdt_tmp, Real rtscavt_sv[]) {
+  // clang-format off
+  //-----------------------------------------------------------------------
+  // resuspension goes to coarse mode
+  //-----------------------------------------------------------------------
+  /*
+  in :: ncol, mm
+  in :: update_dqdt  ! if update dqdt_tmp with rtscavt_sv
+  in :: rcscavt      ! resuspention from convective [kg/kg/s]
+  in :: rsscavt      ! resuspention from stratiform [kg/kg/s]
+
+  inout :: dqdt_tmp ! temporary array to hold tendency for the "current" aerosol species [kg/kg/s]
+  inout :: rtscavt_sv ! resuspension that goes to coarse mode [kg/kg/s]
+  */
+  // clang-format on
+
+  const int mmtoo = aero_model::mmtoo_prevap_resusp(mm);
+
+  // first deduct the current resuspension from the dqdt_tmp of the current
+  // species
+  dqdt_tmp -= (rcscavt + rsscavt);
+
+  // then add the current resuspension to the rtscavt_sv of the appropriate
+  // coarse mode species
+  if (mmtoo > 0)
+    rtscavt_sv[mmtoo] += (rcscavt + rsscavt);
+
+  // then add the rtscavt_sv of the current species to the dqdt_tmp
+  // of the current species. This is not called when lphase==2
+  // note that for so4_a3 and mam3, the rtscavt_sv at this point will have
+  //  resuspension contributions from so4_a1/2/3 and so4c1/2/3
+  if (update_dqdt)
+    dqdt_tmp += rtscavt_sv[mm];
+}
+// =============================================================================
+KOKKOS_INLINE_FUNCTION
+Real calc_sfc_flux(
+    const ThreadTeam &team,
+    Kokkos::View<Real[1], Kokkos::MemoryTraits<Kokkos::Atomic>> scratch,
+    const Real layer_tend, const Real pdel) {
+  // clang-format off
+  // -----------------------------------------------------------------------
+  //  calculate surface fluxes of wet deposition from vertical integration of tendencies
+  // -----------------------------------------------------------------------
+  /*
+  in :: pdel       ! pressure difference between two layers [Pa]
+  in :: layer_tend ! physical tendencies in each layer [kg/kg/s]
+  out :: sflx      ! integrated surface fluxes [kg/m2/s]
+  */
+  // clang-format on
+  const Real gravit = Constants::gravity;
+
+  scratch[0] = 0;
+  team.team_barrier();
+  scratch[0] += layer_tend * pdel / gravit;
```

Comment:
Is every thread accumulating a contribution to `scratch`? Is that why it's `Atomic`?

Reply:
Yes.  This is the problem of a parallel_for over the layers of a column but needing to stop right in the middle of a triple  loop and integrate some value over the column.  Since we are already in a team parallel loop the best I could think of was for each thread to add to a Kokkos::View since the View is shared by all threads.  This requires an atomic_add.  Then a barrier() to make sure all threads are done before extracting the value out of the View so each thread can then continue.  I saw a Kokkos note that low-thread-count contention is not too bad for atomic operations and on a GPU the barrier() should be fast since all threads march in lockstep for the most part.   The other option I tried was a View of length the number of levels so each thread could put the value in the thread slot, then a barrier(), then integrate the View in each thread.   A problem with both is the barrier() which could hang if not all threads reach this point, like if the thread team is not equal to the number of levels.   I'm open to suggestions....

Reply:
It would be ugly to break up into two loops since it is so deep down and all of the local values would be lost.  Convproc is even worse since it needs to do a column integration several functions down from compute_tendencies.